lib/libmbedtls/core/tee_lmd_provider.c

```diff
@@ -2421,72 +2424,263 @@ TEE_Result crypto_mac_final(void *ctx, uint32_t algo, uint8_t *digest,
 /******************************************************************************
  * Authenticated encryption
  ******************************************************************************/
+#define MBEDTLS_CCM_ENCRYPT		0
+#define MBEDTLS_CCM_DECRYPT		1
+#define TEE_CCM_KEY_MAX_LENGTH		32
+#define TEE_CCM_NONCE_MAX_LENGTH	13
+#define TEE_CCM_TAG_MAX_LENGTH		16
+#define TEE_xCM_TAG_MAX_LENGTH		16
+
 #if defined(CFG_CRYPTO_CCM)
-TEE_Result crypto_aes_ccm_alloc_ctx(void **ctx_ret __unused)
-{
-	return TEE_ERROR_NOT_IMPLEMENTED;
+struct tee_ccm_state {
+	mbedtls_ccm_context ctx;	/* the ccm state as defined by LMD */
+	size_t tag_len;			/* tag length */
+	unsigned char iv[16];		/* IV */
+	size_t iv_size;			/*
+					 * IV size in bytes (for ciphers with
+					 * variable-length IVs)
+					 */
+	bool iv_flag;			/* whether IV has been set */
+	unsigned char *aad;		/* AAD */
+	size_t aad_len;			/* AAD size in bytes */
+	size_t aad_totallen;		/*total aad length */
+	size_t payload_len;		/*src_len*/
+};
+
+TEE_Result crypto_aes_ccm_alloc_ctx(void **ctx_ret)
+{
+	int lmd_res = 0;
+	struct tee_ccm_state *ccm;
+	const mbedtls_cipher_info_t *cipher_info = NULL;
+
+	cipher_info = mbedtls_cipher_info_from_values(MBEDTLS_CIPHER_ID_AES,
+						      128, MBEDTLS_MODE_ECB);
+	if (!cipher_info)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	if (cipher_info->block_size != 16)
```

Comment:
For AES, the block size must be 16. So it will be failed to check parameter if it's not 16.
The line is came from `mbedtls_ccm_setkey()`, I think we need to add it here.

Reply:
Yes, you are right. I will add an `assert()` here.